### PR TITLE
fix(atenlib): make functions with dtype attributes trace_only

### LINKED ISSUE #419: Make functions with if and dtype attributes trace_only
According to Rama, if branches needs to produce the same type and this conflicts with some logic we have in the atenlib

### PR DESCRIPTION
**Pull request by justinchuby**
ONNX requires if branches to produce the same type. This requirement conflicts with some logic we have in the atenlib. This PR fixes the logic.

cc @fatcat-z @xiaowuhu @titaiwangms @BowenBao @gramalingam 

Fixes #419

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 1]
```diff
@@ -247,9 +247,6 @@ def _where_input_wrangler(
     "acosh": core_ops.aten_acosh,
     "add": core_ops.aten_add,
     "addmm": core_ops.aten_addmm,
-    "arange_start_step": core_ops.aten_arange_start_step,
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: The issue was not raised by unit tests earlier. Is it because we are using reference evaluator? Is it a bug in reference evaluator that it overlooks the if-branch-same-dtype requirement?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: onnx.checker.check_function missed to check it I think.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 2]
```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
...
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        zero = op.Constant(value_float=0.0)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(zero, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
+            result = op.CastLike(result, end_)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:308]: May be it would be better to handle the `dtype == -1` case separately, since it seems doable without the final cast (eg., from float to some int-type). It is almost the same as the then-branch, except we need `CastLike(_, end_)`.

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:308]: there are types Range doesn't support, commented in code

```
        # Cast input to float if dtype is not supported by Range,
        # because the input dtype may be e.g. bool
        # which Range does not support. The output type is ensured because the output
        # is casted to the specified dtype.
```

Does that sound reasonable?

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 3]
```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
...
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        zero = op.Cast(0, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(zero, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:297]: Is this used, eg., for 16-bit floats? Just wondering, since bool seems like a very special odd-case, though I guess a user could use it.

Comment (gramalingam) [reply onnxscript/function_libs/torch_aten/ops/core.py:297]: Hopefully not for complex, which is tricky

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:297]: Not complex (and not bool actually):

image

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:297]: uint8, int8, float16, bfloat16

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 4]
```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
...
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
+            result = op.CastLike(result, end_)
 
     return result
 
 
-@torch_op("aten::arange", overload=True)
+@torch_op("aten::arange", overload=True, trace_only=True)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:313]: (Not part of this PR, but wondering): What does the name `TRealUnlessFloat16OrInt8` mean? The name seems unconnected to the actual definition.

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: Filed https://github.com/microsoft/onnx-script/issues/423

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:313]: `DOUBLE, FLOAT, INT16, INT32, INT64`. I think it's accurate? Maybe I am missing something?

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 2 above)
Comment (gramalingam) [reply onnxscript/function_libs/torch_aten/ops/core.py:308]: Yes. But in the case where `end` has a supported type and `dtype == -1` (which may be a common case) you will end up with more convoluted code with extra casts. So, I should refine my suggestion to this case.

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: the evaluator actually executes line by line so its behavior may be different when executing a function. @gramalingam should we allow a function mode? e.g. `ORTEvaluator(compile_functions=True)` But that will not allow break points any more...

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: We do run both the reference and ort in test.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:308]: But we don't know if end has a supported type I don't think?

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:308]: Unless we create an overload and dispatch from the exporter

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (gramalingam) [reply onnxscript/function_libs/torch_aten/ops/core.py:313]: Ok. There's bfloat16 missing in the name. And the word `Unless` is misleading. Perhaps it is supposed to be `Less` or `Minus` or `ButNot` or `AndNot`.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 5]
```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
...
+        # which Range does not support. The output type is ensured because the output
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        start = op.Cast(start, to=FLOAT.dtype)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(start, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:339]: In this case, it looks like all input types are supported by range. Special-casing for `dtype == -1` may be useful.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 6]
```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
...
+            result = op.CastLike(result, end_)
 
     return result
 
 
-@torch_op("aten::arange", overload=True)
+@torch_op("aten::arange", overload=True, trace_only=True)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:345]: The function name and parameter name don't match. Is it `step` or `end` ?

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:308]: I thought about this again. You are right. I will update later.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:345]: It's step from the aten signature: arange.start_step

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (gramalingam) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: Yes, as Justin says, execution of control-flow constructs uses python (for better debugging) and will not invoke the ONNX checker (for an IF or LOOP). 

One possibility is to add a checker API, that looks like a normal eager-mode call, but invokes the checker (with an appropriate context). Sort of similar to Justin's suggestion above, except it is just a checker with no-execution.

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:250]: Yes I don't see that as mutually exclusive (do checker first, then eager execute). Maybe a performance concern?

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Done.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:312]: How about `TRealNotFloat16OrBFloat16OrInt8`

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Done

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:318]: Created https://github.com/microsoft/onnx-script/issues/425 to track

**Comment by justinchuby** [comment]
I am going to test this fix with ORT.

**Comment by titaiwangms** [comment]
What happened to CI / test (ubuntu-latest, py310-onnx-weekly) (pull_request) ?

**Comment by justinchuby** [comment]
onnx weekly didn't upload a linux wheel last week. we expect this to be fixed next week.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 7]
```diff
@@ -275,44 +275,77 @@ def aten_any(self: TensorType) -> TensorType:
...
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(start, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if dtype == -1:
+        one = op.CastLike(1.0, end)
+        result = op.Range(start, end, one)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:329]: I assume that inputs start and end are required to be of same type?

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:329]: Yes. We express that in the function signature (mostly true for all functions in atenlib)

**Review by gramalingam — APPROVED** [review]

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (gramalingam) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:255]: Agree. The infrastructure should enable the users to decide whether they want separate tests or always run the checker in eager-mode, etc.
